src/core/service.c

```diff
@@ -1440,6 +1440,93 @@ static bool service_exec_needs_notify_socket(Service *s, ExecFlags flags) {
         return s->notify_access != NOTIFY_NONE;
 }
 
+static int service_create_monitor_md_env(Job *j, char **ret) {
+        _cleanup_free_ char *var = NULL;
+        const char *list_delim = ";";
+        bool first = true;
+        Unit *tu;
+
+        assert(j);
+        assert(ret);
+
+        /* Create an environment variable 'MONITOR_METADATA', if creation is successful
+         * a pointer to it is returned via ret.
+         *
+         * This variable contains a space separated set of fields which relate to
+         * the service(s) which triggered job 'j'. Job 'j' is the JOB_START job for
+         * an OnFailure= or OnSuccess= dependency. Format of the MONITOR_METADATA
+         * variable is as follows:
+         *
+         * MONITOR_METADATA="SERVICE_RESULT=<result-string0>,EXIT_CODE=<exit-code0>,EXIT_STATUS=<exit-status0>,
+         *                   INVOCATION_ID=<id>,UNIT=<triggering-unit0.service>;
+         *                   SERVICE_RESULT=<result-stringN>,EXIT_CODE=<exit-codeN>,EXIT_STATUS=<exit-statusN>,
+         *                   INVOCATION_ID=<id>,UNIT=<triggering-unitN.service>"
```

Comment:
i am not convinced at all about this way to encode stuff. People probably want to use this in shell scripts, but this not really consumable by shell scrpts...

Quite frankly, i'd dumb this down. I mean, yes a failure unit could be started by multiple units, and that's fine, but I think we simply should not pass the metdata in that case where it is not unique.

instead: if there is anything not unique, let's just not pass metadata, let's only pass it if a unit got started as OnFailure=/OnSuccess= of exactly one other unit.

Reply:
The failure unit being started by multiple units is a hard requirement for a reactive monitoring system, we can't lose data.

It can be consumed by scripts, see the unit test:

https://github.com/systemd/systemd/blob/main/test/units/testsuite-68.sh#L79

but if you have a suggestion for a better format I can implement it - this already shipped in a release though, but we might get away with it

Reply:
> The failure unit being started by multiple units is a hard requirement for a reactive monitoring system, we can't lose data.

I don't follow? why is it a hard requirement to have a single unit that is triggered for a number of other units? why not just instantiate a small unit for each failing service? I mean, we have template units for a reason. i.e. `OnFailure=myfailurehandler@%n.service` is kinda how this is supposed to be used.

Reply:
It needs to work for instances too - wasn't the infinite recursion issue fixed last year by making sure it can't happen in these cases? IE, the specifier doesn't get expanded past the first level

Reply:
Well, yes and no. Should it work on instance units? yes, of course. Should it work recursively without limit? i.e. should you be able to have a monitor service for a monitor service for a monitor service. Nope. Certainly not.

My assumption is this: if you have a service called foo.service, then you'd do OnFailure=myfailurehandler@%p.service, but if you have quux@.service then you probably should do OnFailure=myfailurehandler@%p-%i.service or so. i.e. it's OK to tweak the instance to use.

I really don#t think it's a worthy goal to have a single service be executed for failures of a multitude of other services, because these other services will likely fail one by one, and not all atomically at the same moment. But only if they all share the exact same OnFailure= service then you'll likely get that service started exactly once, and it will process the first failure, but the rest will be ignored, because then the failure service is already running, and doesn't get started again. If you however use instances, then the problem goes away: each failing service will get its own failure service instance spawned which can process what is needed and then exit when its done: events will not be lost.